cmd/kata_mdev_plugin/kata_mdev_plugin.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"path"
+	"regexp"
+	"strings"
+	"time"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/debug"
+	"github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+)
+
+const (
+	vfioDevicePath      = "/dev/vfio"
+	iommuGroupDirectory = "/sys/kernel/iommu_groups"
+	iommuGroupDevices   = "devices"
+
+	// Device plugin settings.
+	namespace  = "kata.intel.com"
+	deviceType = "mdev"
+)
+
+type devicePlugin struct {
+	iommuGroupDir string
+	maxDevices    int
+}
+
+func newDevicePlugin(iommuGroupDir string, maxDevices int) *devicePlugin {
+	return &devicePlugin{
+		iommuGroupDir: iommuGroupDir,
+		maxDevices:    maxDevices,
+	}
+}
+
+func (dp *devicePlugin) Scan(notifier deviceplugin.Notifier) error {
+	for {
+		devTree, err := dp.scan()
+		if err != nil {
+			return err
+		}
+
+		notifier.Notify(devTree)
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (dp *devicePlugin) getMdevDeviceSpecs(devDir string, vfioNode string) ([]pluginapi.DeviceSpec, error) {
+
+	UUID := devDir
+	mdevDev := vfioNode
+
+	fmt.Printf("%s device: corresponding MDEV device detected is %s\n", UUID, mdevDev)
+
+	return []pluginapi.DeviceSpec{
+		{
+			HostPath:      mdevDev,
+			ContainerPath: mdevDev,
+			Permissions:   "rw",
+		},
+	}, nil
+
+	return nil, errors.New("Unknown MDEV driver")
+}
+
+func (dp *devicePlugin) scan() (deviceplugin.DeviceTree, error) {
+	devTree := deviceplugin.NewDeviceTree()
+	n := 0
+
+	vfioNodes, err := ioutil.ReadDir(vfioDevicePath)
+
+	if err != nil {
+		fmt.Println("Can't read vfio Device Path with MDEV driver: Skipping")
+	}
+	for _, node := range vfioNodes {
+		if strings.HasSuffix(node.Name(), "vfio") {
+			continue
+		}
+		iommuGroupNode := path.Join(iommuGroupDirectory, node.Name())
+
+		vfioDevices, err2 := ioutil.ReadDir(path.Join(iommuGroupNode, iommuGroupDevices))
+
+		if err2 != nil {
+			return nil, err2
+		}
+		for _, device := range vfioDevices {
+			match, _ := regexp.MatchString("[0-9a-z]{8}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{12}$", device.Name()) // Judge whether it is MDEV decice
```

Comment:
No need to parse this regexp on every iteration, better move this to contants.

Reply:
OK, have moved the parse to contants.